67_RayQueryGeometry/app_resources/render.comp.hlsl

```diff
@@ -25,69 +25,64 @@ float3 unpackNormals3x10(uint32_t v)
+        indices[2] = indices[0] + 2;
+    }
+    else {
+        switch (indexType)
         {
-            indices[0] = primID * 3;
-            indices[1] = indices[0] + 1;
-            indices[2] = indices[0] + 2;
+            case 0: // EIT_16BIT
+                indices = uint32_t3((nbl::hlsl::bda::__ptr<uint16_t3>::create(indexBufferAddress)+primID).deref().load());
+                break;
+            case 1: // EIT_32BIT
+                indices = uint32_t3((nbl::hlsl::bda::__ptr<uint32_t3>::create(indexBufferAddress)+primID).deref().load());
+                break;
         }
     }
 
+    if (normalBufferAddress == 0)
+    {
+        float3 v0 = vk::RawBufferLoad<float3>(vertexBufferAddress + indices[0] * 12);
+        float3 v1 = vk::RawBufferLoad<float3>(vertexBufferAddress + indices[1] * 12);
+        float3 v2 = vk::RawBufferLoad<float3>(vertexBufferAddress + indices[2] * 12);
+
+        return normalize(cross(v2 - v0, v1 - v0));
+    }
+
     float3 n0, n1, n2;
-    switch (instID)
+    switch (normalType)
     {
-        case OT_CUBE:
+        case NT_R8G8B8A8_SNORM:
         {
-            // TODO: document why the alignment is 2 here and nowhere else? isnt the `vertexStride` aligned to more than 2 anyway?
-            uint32_t v0 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[0] * vertexStride, 2u);
-            uint32_t v1 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[1] * vertexStride, 2u);
-            uint32_t v2 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[2] * vertexStride, 2u);
+            uint32_t v0 = vk::RawBufferLoad<uint32_t>(normalBufferAddress + indices[0] * 4);
+            uint32_t v1 = vk::RawBufferLoad<uint32_t>(normalBufferAddress + indices[1] * 4);
+            uint32_t v2 = vk::RawBufferLoad<uint32_t>(normalBufferAddress + indices[2] * 4);
 
             n0 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v0).xyz);
             n1 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v1).xyz);
             n2 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v2).xyz);
         }
         break;
-        case OT_SPHERE:
-        case OT_CYLINDER:
-        case OT_ARROW:
-        case OT_CONE:
+        case NT_R32G32B32_SFLOAT:
         {
-            uint32_t v0 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[0] * vertexStride);
-            uint32_t v1 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[1] * vertexStride);
-            uint32_t v2 = vk::RawBufferLoad<uint32_t>(vertexBufferAddress + indices[2] * vertexStride);
-
-            n0 = normalize(unpackNormals3x10(v0));
-            n1 = normalize(unpackNormals3x10(v1));
-            n2 = normalize(unpackNormals3x10(v2));
+            n0 = normalize(vk::RawBufferLoad<float3>(normalBufferAddress + indices[0] * 12));
+            n1 = normalize(vk::RawBufferLoad<float3>(normalBufferAddress + indices[1] * 12));
+            n2 = normalize(vk::RawBufferLoad<float3>(normalBufferAddress + indices[2] * 12));
```

Comment:
normalization could have been hosted outside the switch to happen later, but oh well